### PR TITLE
omkafka: added implstats counters for librdkafka returned statistics …

### PR DESCRIPTION
**Pull request by abwaheed**
…callback and errors

Adds:
-- statscallback counters
-- librdkafka failure and error counters
-- adds acked message counter

### CODE REVIEW

**Comment by rgerhards** [comment]
thx for the PR, much appreciated. Right now doing final touches for closed 8.34 merge window, will do a full review after that.

**Comment by rgerhards** [comment]
It would be great if you could also update the doc. Good entry point is probably here https://github.com/rsyslog/rsyslog-doc/blob/56ff735233f3ecbbea8ebd02d7fbef843aea5a1d/source/configuration/rsyslog_statistic_counter.rst

File: plugins/omkafka/omkafka.c [hunk 1]
```diff
@@ -743,13 +791,157 @@ deliveryCallback(rd_kafka_t __attribute__((unused)) *rk,
...
+
+/**
+ * This function performs a two level search in stats callback json
+ * object. It iterates over broker objects and for each broker object
+ * returns desired level2 value (such as avg/min/max) for specified
+ * level1 window statistic (such as rtt/throttle/int_latency). Threshold
+ * allows skipping values that are too small, so that they don't
+ * impact on aggregate averaged value that is returned.
+ */
+static uint64 jsonExtractWindoStats(struct fjson_object * stats_object,
```
Comment (rgerhards) [inline plugins/omkafka/omkafka.c]: we indent by tabs. Please change (one tab = 8 chars standard width)

File: plugins/omkafka/omkafka.c [hunk 2]
```diff
@@ -581,8 +622,8 @@ writeKafka(instanceData *const pData, uchar *const msg,
 	pthread_rwlock_t *dynTopicLock = NULL;
 	failedmsg_entry* fmsgEntry;
 	int topic_mut_locked = 0;
-#if RD_KAFKA_VERSION >= 0x00090400
```
Comment (rgerhards) [inline plugins/omkafka/omkafka.c:584]: Is this really correct? The #if seems to guard the kafka type, doen'st make so much sense to me to guard the int64_t. Or am I overlooking something?

File: plugins/omkafka/omkafka.c (hunk 1 above)
Comment (rgerhards) [reply plugins/omkafka/omkafka.c]: oh, and we follow the Gnu rule of starting the function name on column 1 of the next line

File: plugins/omkafka/omkafka.c (hunk 2 above)
Comment (abwaheed) [reply plugins/omkafka/omkafka.c:584]: I only pulled rd_kafka_resp_err_t msg_kafka_response; from under #if as I can use it for both librdkafka version types. int64_t ttMsgTimestamp was also already there and is being used to pass timestamp of resubmit to rd_kafka_producev(). Interface for older version looks different, which is probably why it was conditionally declared for newer versions.

**Comment by rgerhards** [comment]
doc at https://github.com/rsyslog/rsyslog-doc/pull/640

**Comment by rgerhards** [comment]
ci had unrelated issues, restarting (may take a while due to busy CI servers)

File: plugins/omkafka/omkafka.c [hunk 3]
```diff
@@ -743,13 +786,156 @@ deliveryCallback(rd_kafka_t __attribute__((unused)) *rk,
...
 }
 
+/**
+ * This function looks for a json object that corresponds to the
+ * passed name and returns it is found. Otherwise returns NULL.
+ * It will be used for processing stats callback json object.
+ */
+static struct fjson_object * get_object(struct fjson_object *fj_obj,
+		const char * name) {
+  struct fjson_object_iterator it = fjson_object_iter_begin(fj_obj);
```
Comment (rgerhards) [inline plugins/omkafka/omkafka.c]: I guess you missed some indention here - sorry to be a pest

File: plugins/omkafka/omkafka.c [hunk 4]
```diff
@@ -743,13 +786,156 @@ deliveryCallback(rd_kafka_t __attribute__((unused)) *rk,
...
+ *
+ * Here we have extracted windows stats: rtt, throttle time, and internal
+ * latency averages. These values will be logged as impstats messages.
+ */
+static int
+statsCallback(rd_kafka_t __attribute__((unused)) *rk,
+		char *json, size_t __attribute__((unused)) json_len,
+		void __attribute__((unused)) *opaque) {
+	char buf[2048];
+    char handler_name[1024] = "unknown";
```
Comment (rgerhards) [inline plugins/omkafka/omkafka.c]: and indention here as well

File: plugins/omkafka/omkafka.c (hunk 3 above)
Comment (abwaheed) [reply plugins/omkafka/omkafka.c]: sorry, missed that in first round. I'll update this and check others as well.

**Comment by rgerhards** [comment]
doc PR: https://github.com/rsyslog/rsyslog-doc/pull/640

**Comment by rgerhards** [comment]
merge pr #2598 went through fine, so finally merged. #2598 will be closed unmerged (changes were only housekeeping).

**Pull request merged**